### PR TITLE
Player class task7

### PR DESCRIPTION
**Pull request by BarTes8**

### CODE REVIEW

File: shm/inc/Player.hpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+#include "Cargo.hpp"
+#include "Ship.hpp"
```
Comment (MoreThanADC) [inline shm/inc/Player.hpp]: Do relative path like "shm/inc/Cargo.hpp"  etc.

File: shm/inc/Ship.hpp [hunk 2]
```diff
@@ -1,5 +1,8 @@
 #pragma once
+#include "../inc/Cargo.hpp"
```
Comment (MoreThanADC) [inline shm/inc/Ship.hpp]: Do relative path like "shm/inc/Cargo.hpp"

File: shm/main.cpp [hunk 3]
```diff
@@ -1,3 +1,4 @@
+#include "inc/Player.hpp"
```
Comment (MoreThanADC) [inline shm/main.cpp:1]: Do relative path

File: shm/src/Player.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#include "../inc/Player.hpp"
+
+Player::Player(std::shared_ptr<Ship> ship, size_t money, size_t availableSpace)
+        : ship_(std::make_shared<Ship>(ship)), money_(money), availableSpace_(availableSpace)
+    {}
```
Comment (lukasz-kukulka) [inline shm/src/Player.cpp]: Every argument from initialization list in new line will be more clear

**Review by lukasz-kukulka — APPROVED** [review]

**Review by mariuszlisowski — APPROVED** [review]

File: shm/src/Player.cpp [hunk 5]
```diff
@@ -1,7 +1,7 @@
 #include "../inc/Player.hpp"
 
 Player::Player(std::shared_ptr<Ship> ship, size_t money, size_t availableSpace)
```
Comment (mariuszlisowski) [inline shm/src/Player.cpp]: unique_ptr

**Review by mariuszlisowski — CHANGES_REQUESTED** [review]

File: shm/src/Player.cpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
+#include "inc/Player.hpp"
```
Comment (mariuszlisowski) [inline shm/src/Player.cpp]: "shm/inc/Player.hpp"

File: shm/src/Cargo.cpp [hunk 7]
```diff
@@ -1,4 +1,4 @@
-#include "Cargo.hpp"
+#include "inc/Cargo.hpp"
```
Comment (mariuszlisowski) [inline shm/src/Cargo.cpp]: "shm/inc/Cargo.hpp"

File: shm/inc/Player.hpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include "inc/Cargo.hpp"
+#include "inc/Ship.hpp"
```
Comment (mariuszlisowski) [inline shm/inc/Player.hpp]: "shm/inc/Cargo.hpp"
"shm/inc/Ship.hpp"

File: shm/inc/Ship.hpp [hunk 9]
```diff
@@ -1,5 +1,8 @@
 #pragma once
+#include "inc/Cargo.hpp"
```
Comment (mariuszlisowski) [inline shm/inc/Ship.hpp]: "shm/inc/Cargo.hpp"

File: shm/inc/Player.hpp [hunk 10]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include "inc/Cargo.hpp"
```
Comment (mariuszlisowski) [inline shm/inc/Player.hpp]: Line

File: shm/src/Ship.cpp [hunk 11]
```diff
@@ -1,4 +1,4 @@
-#include "../inc/Ship.hpp"
+#include "inc/Ship.hpp"
```
Comment (mariuszlisowski) [inline shm/src/Ship.cpp]: "shm/inc/Ship.hpp"

**Review by mariuszlisowski — CHANGES_REQUESTED** [review]

File: shm/inc/Ship.hpp [hunk 12]
```diff
@@ -15,6 +18,9 @@ class Ship {
     size_t getSpeed() const { return speed_; };
```
Comment (mariuszlisowski) [inline shm/inc/Ship.hpp:20]: return int

File: shm/src/Player.cpp [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    if (ship_) {
+        for (size_t i = 0; i < ship_->getHowManyCargos(); i++) {
+            availableSpace -= ship_->getCargo(i)->getAmount();
+        }
+    }
+    return availableSpace;
+}*/
+
+size_t Player::countAvailableSpace() const {
+    size_t usedSpace = std::accumulate(ship_->getVectorCargo().cbegin(), ship_->getVectorCargo().cend(), 0);
```
Comment (mariuszlisowski) [inline shm/src/Player.cpp]: Iteration through a vector of shared pointers but int type needed to add up (amount of cargo).
Thus error: invalid operands to binary expression ('int' and 'const std::__1::shared_ptr<Cargo>')
Hint: each vector element contains a cargo which needs to be accessed.

std::accumulate returns int type but size_t used (which is good) - make accumulate to return size_t

**Review by mariuszlisowski — CHANGES_REQUESTED** [review]

File: shm/src/Player.cpp [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    return availableSpace;
+}*/
+
+size_t Player::countAvailableSpace() const {
+    size_t usedSpace = 0;
+    /*size_t usedSpace = std::accumulate(ship_->getVectorCargo().cbegin(), ship_->getVectorCargo().cend(), 0),
+        [&](auto& rhs, auto& lhs){ return (*rhs + *lhs); };*/
+    std::for_each(ship_->getVectorCargo().cbegin(), ship_->getVectorCargo().cend(), 
+        [&usedSpace](const auto& ptrCargo){ usedSpace += *ptrCargo; });
```
Comment (mariuszlisowski) [inline shm/src/Player.cpp]: Dosen't make sense because usedSpade is size_t type but *ptrCargo i Cargo type - cannot be added.

**Review by mariuszlisowski — CHANGES_REQUESTED** [review]

File: shm/src/Player.cpp [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+size_t Player::countAvailableSpace() const {
+    auto usedSpace =
+    std::accumulate(ship_->getVectorCargo().cbegin(),
+                    ship_->getVectorCargo().cend(),
+                    static_cast<size_t>(0),
+                    [](size_t sum, const auto& cargo) {
+                        return sum += cargo->getAmount();
+                    });
```
Comment (mariuszlisowski) [inline shm/src/Player.cpp:22]: return available space

File: shm/src/Ship.cpp [hunk 16]
```diff
@@ -1,5 +1,4 @@
-#include "../inc/Ship.hpp"
-#include <iostream>
+#include "shm/inc/Ship.hpp"
 
 Ship::Ship(size_t id, const std::string& name, size_t speed, size_t maxCrew, size_t capacity)
```
Comment (mariuszlisowski) [inline shm/src/Ship.cpp]: int id

File: shm/inc/Ship.hpp [hunk 17]
```diff
@@ -10,11 +15,13 @@ class Ship {
     Ship& operator+=(const size_t);
```
Comment (mariuszlisowski) [inline shm/inc/Ship.hpp]: Ship(int, const std::string&, size_t, size_t, size_t);

**Review by mariuszlisowski — CHANGES_REQUESTED** [review]

**Review by mariuszlisowski — APPROVED** [review]

**Pull request merged**